rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,455 @@
+use crate::errors::SymCryptError;
+use core::ffi::c_void;
+use std::mem;
+use std::pin::Pin;
+use std::ptr;
+use std::vec;
+use symcrypt_sys;
+
+pub struct GcmState {
+    inner: Pin<Box<GcmStateInner>>,
+}
+
+struct GcmStateInner {
+    state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+}
+
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime.
+/// Using an inner GCM state that is Pin<Box<T>> since the memory address for Self is moved around when returning from GcmState::new()
+/// My guess is that it has to do with Result<> wrapping the Self and SymCryptError.
+///
+/// auth_part, encrypt_part, decrypt_part take in an allocated buffer as an out parameter for performance reasons. This is for scenarios
+/// such as encrypting over a stream of data; allocating and copying data from a return will be costly performance wise.
+///
+/// Since auth_part, encrypt_part and decrypt_part may be called multiple times, you must call encrypt_final/decrypt_final at the end to ensure
+/// that the encryption/decryption has completed successfully.
+///
+/// The only accepted Cipher for GCM is AesBlock.
+impl GcmState {
+    pub fn new(
+        key: &[u8],
+        nonce: &[u8; 12], // GCM nonce length must be 12 bytes
+        cipher: BlockCipherType,
+    ) -> Result<Self, SymCryptError> {
+        let mut instance = GcmState {
+            inner: Box::pin(GcmStateInner {
+                state: symcrypt_sys::SYMCRYPT_GCM_STATE::default(),
+                expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default(),
+            }),
+        };
+
+        gcm_expand_key(
+            key,
+            &mut instance.inner.expanded_key,
+            convert_cipher(cipher),
+        )?;
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptGcmInit(
+                &mut instance.inner.state,
+                &instance.inner.expanded_key,
+                nonce.as_ptr(),
+                nonce.len() as symcrypt_sys::SIZE_T,
+            );
+            Ok(instance)
+        }
+    }
+
+    // Can be called multiple times
+    pub fn auth_part(&mut self, auth_data: &[u8]) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptGcmAuthPart(
+                &mut self.inner.state,
+                auth_data.as_ptr(),
+                auth_data.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    // Can be called multiple times
+    // Takes in a buffer as an out parameter for performance reasons
+    pub fn encrypt_part(&mut self, plain_text: &[u8], cipher_text_buffer: &mut [u8]) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptGcmEncryptPart(
+                &mut self.inner.state,
+                plain_text.as_ptr(),
+                cipher_text_buffer.as_mut_ptr(),
+                cipher_text_buffer.len() as symcrypt_sys::SIZE_T,
```

Comment:
>cipher_text_buffer.len() as symcrypt_sys::SIZE_T, [](http://example.com/codeflow?start=16&length=49)

Need to check that plain_text.len() == cipher_text_buffer.len()

Reply:
Instead of returning an error, might be OK to panic instead. It is a caller mistake if they aren't ==

Question for Eric?